File: 48-rotate-image/48-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
    
        (0,1) <-- (2,0)
        (2,0) <-- (3,2)
        (3,2) <-- (1,3)
        (1,3) <-- (0,1)

        """

        n = len(matrix)
        for i in range(n//2 + n%2):
            for j in range(n//2):
                temp = matrix[i][j]
                matrix[i][j] = matrix[n-1-j][i]
                matrix[n-1-j][i] = matrix[n-1-i][n-1-j]
                matrix[n-1-i][n-1-j] = matrix[j][n-1-i]
                matrix[j][n-1-i] = temp

File: 48-rotate-image/48-rotate-image.c
void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void rotate(int** matrix, int matrixSize, int* matrixColSize){
    int temp, tmp;
    int r,c;
    int i,j;
    
    int **visited;
    visited = calloc(matrixSize, sizeof(int *));
    if (visited == NULL) {
        return;
    }
    
    for (i = 0; i < matrixSize; i++) {
        visited[i] = calloc(matrixColSize[0], sizeof(int));
        if (visited[i] == NULL) {
            return;
        }
    }

    for (i = 0; i < matrixSize/2; i++) {
        for (j = 0; j < matrixColSize[i]-i; j++) {
            r = i;
            c = j;

            temp = matrix[r][c];
            while (visited[r][c] == 0) {
                swap(&matrix[c][matrixColSize[0]-1-r], &temp);
                visited[r][c] = 1;

                tmp = c;
                c = matrixColSize[0]-1-r;
                r = tmp;
            }
        }   
    }
}